### PR TITLE
Fix missing paths when using typescript's module nodenext feature

### PR DESCRIPTION
**Pull request by IgnusG**
This resolution requires relative/mapped import paths to end with .js
This PR detects if a path ends with `.js` and if so adds more extension type paths to try, that omit this extension when attaching additional extensions to match the actual existing files.

Reference https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/#esm-nodejs

Should resolve https://github.com/dividab/tsconfig-paths/pull/213 (edit 1: wrong issue sorry, edit 2: actually I believe this **might** resolve that issue as well) & https://github.com/aleclarson/vite-tsconfig-paths/issues/59

### CODE REVIEW

**Comment by WoodyWoodsta** [comment]
Is this going anywhere?

**Review by aleclarson — APPROVED** [review]

**Comment by aleclarson** [comment]
Can you copy over the test from https://github.com/dividab/tsconfig-paths/pull/213 ?

**Comment by WoodyWoodsta** [comment]
I would also like to know if this handles `.mjs` files, which is what is emitted by the TS compiler when files are `.mts` files.

**Comment by aleclarson** [comment]
> I would also like to know if this handles `.mjs` files, which is what is emitted by the TS compiler when files are `.mts` files.

Good catch. There should definitely be a test for that.

**Comment by IgnusG** [comment]
I'll add both tests and check the mjs support 👍

File: src/__tests__/try-path.test.ts [hunk 1]
```diff
@@ -95,6 +95,58 @@ describe("mapping-entry", () => {
...
+          type: "extension",
+          path: join("/absolute", "base", "url", "foo2", "bar.tsx"),
+        },
+      ])
+    );
+  });
+
+  it("should include paths with ending .mjs removed that matches requested module", () => {
+    const result = getPathsToTry(
+      [".ts", ".tsx", ".mts", ".mtsx"],
```
Comment (WoodyWoodsta) [inline src/__tests__/try-path.test.ts:126]: So I'm not sure `.mtsx` is a file extension that exists - not to the best of my knowledge anyway. What _is_ a valid spec extension is `.cts` and `.cjs` (for CommonJS files). See https://nodejs.org/api/esm.html#interoperability-with-commonjs

**Review by WoodyWoodsta — COMMENTED** [review]
I'm coming here from https://github.com/aleclarson/vite-tsconfig-paths/issues/54, so I would need to fiddle around to test it - but will do if I manage to find a gap.

File: src/__tests__/try-path.test.ts (hunk 1 above)
Comment (WoodyWoodsta) [reply src/__tests__/try-path.test.ts:126]: And I use the work "spec" in a sense that this is the `stability: 2` decision made by Node.js.

Comment (aleclarson) [reply src/__tests__/try-path.test.ts:126]: We should cover all the bases that Vite does

https://github.com/vitejs/vite/blob/632fedf87fbcb81b2400571886faf8a8b92376e4/packages/vite/src/node/__tests__/utils.spec.ts#L124-L144

Comment (WoodyWoodsta) [reply src/__tests__/try-path.test.ts:126]: TIL :)

Comment (IgnusG) [reply src/__tests__/try-path.test.ts:126]: > So I'm not sure .mtsx is a file extension that exists

I was just testing you :P (but in reality I just copy pasted the same thing and added an m before it at night)

> We should cover all the bases that Vite does

I'll try to adjust the code to support the ones listed in the linked spec file (probably tomorrow)

**Comment by IgnusG** [comment]
@jonaskello could you please take a look through this when you have the time?

**Review by WoodyWoodsta — APPROVED** [review]

**Comment by IgnusG** [comment]
For anyone wanting to use this with yarn add this resolution to your package.json

```json
resolutions: {
  "tsconfig-paths": "github:ignusg/tsconfig-paths@dev"
}
```

**Comment by datner** [comment]
Hey, this pr was thoroughly tests, reviewed, and even approved!
This is great work and an important step in moving to full esm.

Is there a reason why this isn't merged?

**Comment by jonaskello** [comment]
Yes, seems like this change has enough tests so let's merge it.

**Review by SamyDgz — APPROVED** [review]

**Pull request merged**